src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs

```diff
@@ -96,33 +101,43 @@ public virtual void Invoke(CancellationToken cancellationToken)
 
         public void InvokeCore(Func<Document> getFromDocument, CancellationToken cancellationToken)
         {
+            this.AssertIsForeground();
+
             var extensionManager = this.Workspace.Services.GetService<IExtensionManager>();
             extensionManager.PerformAction(Provider, () =>
             {
-                IEnumerable<CodeActionOperation> operations = null;
-
-                // NOTE: As mentioned above, we want to avoid computing the operations on the UI thread.
-                // However, for CodeActionWithOptions, GetOptions() might involve spinning up a dialog
-                // to compute the options and must be done on the UI thread.
-                var actionWithOptions = this.CodeAction as CodeActionWithOptions;
-                if (actionWithOptions != null)
-                {
-                    var options = actionWithOptions.GetOptions(cancellationToken);
-                    if (options != null)
-                    {
-                        operations = GetOperationsAsync(actionWithOptions, options, cancellationToken).WaitAndGetResult(cancellationToken);
-                    }
-                }
-                else
+                this.WaitIndicator.Wait(CodeAction.Title, CodeAction.Message, allowCancel: true, action: context =>
                 {
-                    operations = GetOperationsAsync(cancellationToken).WaitAndGetResult(cancellationToken);
-                }
+                    InvokeWorker(getFromDocument, cancellationToken);
```

Comment:
Question: Do we need to create a linked cancellation token source here with the cancellation token passed into InvokeCore and given wait context's cancellation token? If some method down the stack checks for `cancellation.ThrowIfCancellationRequested`, will it throw if wait context's cancellation has fired?